swarm/storage/mru/lookup/lookup_test.go

```diff
@@ -0,0 +1,324 @@
+		if data != nil && data.Time <= now {
+			return data, nil
+		}
+		return nil, nil
+	}
+}
+
+func TestLookup(t *testing.T) {
+
+	store := make(Store)
+	readCount := 0
+	readFunc := makeReadFunc(store, &readCount)
+
+	// write an update every month for 12 months 3 years ago and then silence for two years
+
+	now := uint64(1533799046)
+	var epoch lookup.Epoch
+
+	var lastData *Data
+	for i := uint64(0); i < 12; i++ {
+		t := uint64(now - Year*3 + i*Month) // update every month for 12 months 3 years ago and then silence for two years
+		data := Data{
+			Payload: t, //our "payload" will be the timestamp itself.
+			Time:    t,
+		}
+		epoch = update(store, epoch, t, &data)
+		lastData = &data
+	}
+
+	// try to get the last value
+
+	value, err := lookup.Lookup(now, lookup.NoClue, readFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	readCountWithoutHint := readCount
+
+	if value != lastData {
+		t.Fatalf("Expected lookup to return the last written value: %v. Got %v", lastData, value)
+	}
+
+	// reset the read count for the next test
+	readCount = 0
+	// Provide a hint to get a faster lookup. In particular, we give the exact location of the last update
+	value, err = lookup.Lookup(now, epoch, readFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if value != lastData {
+		t.Fatalf("Expected lookup to return the last written value: %v. Got %v", lastData, value)
+	}
+
+	if readCount > readCountWithoutHint {
+		t.Fatalf("Expected lookup to complete with fewer reads than %d since we provided a hint. Did %d reads.", readCountWithoutHint, readCount)
+	}
+
+	// try to get an intermediate value
+	// if we look for a value in now - Year*3 + 6*Month, we should get that value
+	// Since the "payload" is the timestamp itself, we can check this.
+
+	expectedTime := now - Year*3 + 6*Month
+
+	value, err = lookup.Lookup(expectedTime, lookup.NoClue, readFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	data, ok := value.(*Data)
+
+	if !ok {
+		t.Fatal("Expected value to contain data")
+	}
+
+	if data.Time != expectedTime {
+		t.Fatalf("Expected value timestamp to be %d, got %d", data.Time, expectedTime)
+	}
+
+}
+
+func TestOneUpdateAt0(t *testing.T) {
+
+	store := make(Store)
+	readCount := 0
+
+	readFunc := makeReadFunc(store, &readCount)
+	now := uint64(1533903729)
+
+	var epoch lookup.Epoch
+	data := Data{
+		Payload: 79,
+		Time:    0,
+	}
+	update(store, epoch, 0, &data)
+
+	value, err := lookup.Lookup(now, lookup.NoClue, readFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if value != &data {
+		t.Fatalf("Expected lookup to return the last written value: %v. Got %v", data, value)
+	}
+}
+
+func TestLookupFail(t *testing.T) {
+
+	store := make(Store)
+	readCount := 0
+
+	readFunc := makeReadFunc(store, &readCount)
+	now := uint64(1533903729)
+
+	// don't write anything and try to look up.
+	// we're testing we don't get stuck in a loop
+
+	value, err := lookup.Lookup(now, lookup.NoClue, readFunc)
```

Comment:
Do we expect a certain readcount here, and should we check it?

Reply:
Fixed. This helped improve the algorithm too as I was making an extra lookup near t=0.